### PR TITLE
bugfix: correct mouseMove under AzToolsFrameworkHelper

### PR DESCRIPTION
**Pull request by pollend**
REF: https://github.com/o3de/o3de/issues/6481
Signed-off-by: Michael Pollind <mpollind@gmail.com>

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.cpp [hunk 1]
```diff
@@ -45,14 +45,15 @@ namespace UnitTest
     // - https://lists.qt-project.org/pipermail/development/2019-July/036873.html
     void MouseMove(QWidget* widget, const QPoint& initialPositionWidget, const QPoint& mouseDelta, const Qt::MouseButton mouseButton)
     {
-        QPoint nextPosition = widget->mapToGlobal(initialPositionWidget + mouseDelta);
+        QPoint localPosition = initialPositionWidget + mouseDelta;
+        QPoint globalPosition = widget->mapToGlobal(localPosition);
```
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.cpp:49]: uber nit: Could these both be `const` and be called `nextLocalPosition` and `nextGlobalPosition` respectively.

**Review by hultonha — APPROVED** [review]
Change looks good! I think the only thing missing is a test to catch the bug that existed before (should be short to write). Thanks very much for fixing this! :+1:

File: Code/Framework/AzToolsFramework/AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.cpp (hunk 1 above)
Comment (pollend) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.cpp:49]: sure

**Review by amznestebanpapp — APPROVED** [review]

**Comment by amznestebanpapp** [comment]
@pollend the new test is failing https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/PR-6493/1/pipeline

**Comment by pollend** [comment]
> @pollend the new test is failing https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/PR-6493/1/pipeline

that is strange. didn't see this when I ran it locally wonder what I did wrong?

```
[2021-12-22T18:47:58.881Z] [ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/0
[2021-12-22T18:47:58.881Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:56: Failure
[2021-12-22T18:47:58.881Z] The difference between mouseLocalPos.x() and expectedPosition.x() is 2147483658, which exceeds 1.0f, where
[2021-12-22T18:47:58.881Z] mouseLocalPos.x() evaluates to -2147483648,
[2021-12-22T18:47:58.881Z] expectedPosition.x() evaluates to 10, and

```

**Review by cgalvan — APPROVED** [review]

**Comment by cgalvan** [comment]
I've kicked off a new AR run w/ the latest changes

**Comment by cgalvan** [comment]
Mouse move unit tests still fail on AR:

```
[2022-01-03T17:05:01.876Z] [----------] 3 tests from All/MouseMoveAzToolsFrameworkTestHelperFixture

[2022-01-03T17:05:01.876Z] [ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/0

[2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:63: Failure

[2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.x() and expectedPosition.x() is 2147483658, which exceeds 1.0f, where

[2022-01-03T17:05:01.876Z] mouseLocalPos.x() evaluates to -2147483648,

[2022-01-03T17:05:01.876Z] expectedPosition.x() evaluates to 10, and

[2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.

[2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:64: Failure

[2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.y() and expectedPosition.y() is 2147483658, which exceeds 1.0f, where

[2022-01-03T17:05:01.876Z] mouseLocalPos.y() evaluates to -2147483648,

[2022-01-03T17:05:01.876Z] expectedPosition.y() evaluates to 10, and

[2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.

[2022-01-03T17:05:01.876Z] [  FAILED  ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/0, where GetParam() = 32-byte object <00-00 00-00 00-00 00-00 64-00 00-00 64-00 00-00 00-00 00-00 00-00 00-00 0A-00 00-00 0A-00 00-00> (20 ms)

[2022-01-03T17:05:01.876Z] [ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/1

[2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:63: Failure

[2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.x() and expectedPosition.x() is 2147483539, which exceeds 1.0f, where

[2022-01-03T17:05:01.876Z] mouseLocalPos.x() evaluates to 2147483549,

[2022-01-03T17:05:01.876Z] expectedPosition.x() evaluates to 10, and

[2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.

[2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:64: Failure

[2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.y() and expectedPosition.y() is 2147483539, which exceeds 1.0f, where

[2022-01-03T17:05:01.876Z] mouseLocalPos.y() evaluates to 2147483549,

[2022-01-03T17:05:01.876Z] expectedPosition.y() evaluates to 10, and

[2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.

[2022-01-03T17:05:01.876Z] [  FAILED  ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/1, where GetParam() = 32-byte object <64-00 00-00 64-00 00-00 64-00 00-00 64-00 00-00 00-00 00-00 00-00 00-00 0A-00 00-00 0A-00 00-00> (1 ms)

[2022-01-03T17:05:01.876Z] [ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/2

[2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:63: Failure

[2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.x() and expectedPosition.x() is 2147483559, which exceeds 1.0f, where

[2022-01-03T17:05:01.876Z] mouseLocalPos.x() evaluates to 2147483629,

[2022-01-03T17:05:01.876Z] expectedPosition.x() evaluates to 70, and

[2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.

[2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:64: Failure

[2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.y() and expectedPosition.y() is 2147483559, which exceeds 1.0f, where

[2022-01-03T17:05:01.876Z] mouseLocalPos.y() evaluates to 2147483629,

[2022-01-03T17:05:01.876Z] expectedPosition.y() evaluates to 70, and

[2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.

[2022-01-03T17:05:01.876Z] [  FAILED  ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/2, where GetParam() = 32-byte object <14-00 00-00 14-00 00-00 64-00 00-00 64-00 00-00 32-00 00-00 32-00 00-00 14-00 00-00 14-00 00-00> (0 ms)

[2022-01-03T17:05:01.876Z] [----------] 3 tests from All/MouseMoveAzToolsFrameworkTestHelperFixture (21 ms total)
```

**Comment by pollend** [comment]
> Mouse move unit tests still fail on AR:
> 
> ```
> [2022-01-03T17:05:01.876Z] [----------] 3 tests from All/MouseMoveAzToolsFrameworkTestHelperFixture
> 
> [2022-01-03T17:05:01.876Z] [ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/0
> 
> [2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:63: Failure
> 
> [2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.x() and expectedPosition.x() is 2147483658, which exceeds 1.0f, where
> 
> [2022-01-03T17:05:01.876Z] mouseLocalPos.x() evaluates to -2147483648,
> 
> [2022-01-03T17:05:01.876Z] expectedPosition.x() evaluates to 10, and
> 
> [2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.
> 
> [2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:64: Failure
> 
> [2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.y() and expectedPosition.y() is 2147483658, which exceeds 1.0f, where
> 
> [2022-01-03T17:05:01.876Z] mouseLocalPos.y() evaluates to -2147483648,
> 
> [2022-01-03T17:05:01.876Z] expectedPosition.y() evaluates to 10, and
> 
> [2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.
> 
> [2022-01-03T17:05:01.876Z] [  FAILED  ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/0, where GetParam() = 32-byte object <00-00 00-00 00-00 00-00 64-00 00-00 64-00 00-00 00-00 00-00 00-00 00-00 0A-00 00-00 0A-00 00-00> (20 ms)
> 
> [2022-01-03T17:05:01.876Z] [ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/1
> 
> [2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:63: Failure
> 
> [2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.x() and expectedPosition.x() is 2147483539, which exceeds 1.0f, where
> 
> [2022-01-03T17:05:01.876Z] mouseLocalPos.x() evaluates to 2147483549,
> 
> [2022-01-03T17:05:01.876Z] expectedPosition.x() evaluates to 10, and
> 
> [2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.
> 
> [2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:64: Failure
> 
> [2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.y() and expectedPosition.y() is 2147483539, which exceeds 1.0f, where
> 
> [2022-01-03T17:05:01.876Z] mouseLocalPos.y() evaluates to 2147483549,
> 
> [2022-01-03T17:05:01.876Z] expectedPosition.y() evaluates to 10, and
> 
> [2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.
> 
> [2022-01-03T17:05:01.876Z] [  FAILED  ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/1, where GetParam() = 32-byte object <64-00 00-00 64-00 00-00 64-00 00-00 64-00 00-00 00-00 00-00 00-00 00-00 0A-00 00-00 0A-00 00-00> (1 ms)
> 
> [2022-01-03T17:05:01.876Z] [ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/2
> 
> [2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:63: Failure
> 
> [2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.x() and expectedPosition.x() is 2147483559, which exceeds 1.0f, where
> 
> [2022-01-03T17:05:01.876Z] mouseLocalPos.x() evaluates to 2147483629,
> 
> [2022-01-03T17:05:01.876Z] expectedPosition.x() evaluates to 70, and
> 
> [2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.
> 
> [2022-01-03T17:05:01.876Z] /data/workspace/o3de/Code/Framework/AzToolsFramework/Tests/AzToolFrameworkTestHelperTest.cpp:64: Failure
> 
> [2022-01-03T17:05:01.876Z] The difference between mouseLocalPos.y() and expectedPosition.y() is 2147483559, which exceeds 1.0f, where
> 
> [2022-01-03T17:05:01.876Z] mouseLocalPos.y() evaluates to 2147483629,
> 
> [2022-01-03T17:05:01.876Z] expectedPosition.y() evaluates to 70, and
> 
> [2022-01-03T17:05:01.876Z] 1.0f evaluates to 1.
> 
> [2022-01-03T17:05:01.876Z] [  FAILED  ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/2, where GetParam() = 32-byte object <14-00 00-00 14-00 00-00 64-00 00-00 64-00 00-00 32-00 00-00 32-00 00-00 14-00 00-00 14-00 00-00> (0 ms)
> 
> [2022-01-03T17:05:01.876Z] [----------] 3 tests from All/MouseMoveAzToolsFrameworkTestHelperFixture (21 ms total)
> ```

I'm a little stuck here, it passes when I run it locally.

**Comment by hultonha** [comment]
The `2147483658` look suspect (max `int32` + 1), could it be something is getting initialized in debug to zero and not when the tests run on Jenkins?

**Comment by pollend** [comment]
> The `2147483658` look suspect (max `int32` + 1), could it be something is getting initialized in debug to zero and not when the tests run on Jenkins?

strange, since its only the new tests that fail.

**Comment by hultonha** [comment]
> > The `2147483658` look suspect (max `int32` + 1), could it be something is getting initialized in debug to zero and not when the tests run on Jenkins?
> 
> strange, since its only the new tests that fail, probably something small that I'm missing.

@pollend apologies but can you remind me how to grab your branch locally (I remember you mentioning there was a GitHub option to do this). I'll see if I can repro locally too or not. My upstream is the o3de repo but origin is different. I think I should still be able to grab your branch though. Thanks!

**Comment by pollend** [comment]
> > > The `2147483658` look suspect (max `int32` + 1), could it be something is getting initialized in debug to zero and not when the tests run on Jenkins?
> > 
> > strange, since its only the new tests that fail, probably something small that I'm missing.
> 
> @pollend apologies but can you remind me how to grab your branch locally (I remember you mentioning there was a GitHub option to do this). I'll see if I can repro locally too or not. My upstream is the o3de repo but origin is different. I think I should still be able to grab your branch though. Thanks!

You can use GitHub CLI. https://cli.github.com/manual/gh_pr_checkout

**Comment by hultonha** [comment]
> > > > The `2147483658` look suspect (max `int32` + 1), could it be something is getting initialized in debug to zero and not when the tests run on Jenkins?
> > > 
> > > 
> > > strange, since its only the new tests that fail, probably something small that I'm missing.
> > 
> > 
> > @pollend apologies but can you remind me how to grab your branch locally (I remember you mentioning there was a GitHub option to do this). I'll see if I can repro locally too or not. My upstream is the o3de repo but origin is different. I think I should still be able to grab your branch though. Thanks!
> 
> You can use GitHub CLI. https://cli.github.com/manual/gh_pr_checkout

That was it! 😅 Thank you! Will try it out later this afternoon, thanks! 😄

**Comment by hultonha** [comment]
Oh wait I think I just realized why it might be! Do you have the setting set locally in your settings registry? For enabling wraparound? It's probably not working on the build machines because they won't have that setting enabled 😬 So the tests are failing, does that sound plausible? (I've just synced your branch and am testing to see if that's the case too)

**Comment by hultonha** [comment]
One solution is to make an interface for checking if wraparound is enabled or not (so we can cutout relying on the Settings Registry entirely) or we need a way to modify that setting from the test by creating a settings registry instance in the fixture (this is likely the less optimal solution but might be less code to change to fix it...) I'll have a little look and see

**Comment by hultonha** [comment]
Argh never mind it wasn't that 😞 I was thinking of the other wraparound change (sorry still getting caught up after being out for while 🙈) The test does pass for me locally too 🤔 

```
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from All/MouseMoveAzToolsFrameworkTestHelperFixture
[ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/0
[       OK ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/0 (3 ms)
[ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/1
[       OK ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/1 (0 ms)
[ RUN      ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/2
[       OK ] All/MouseMoveAzToolsFrameworkTestHelperFixture.MouseMove_AzToolsFrameworkTestHelpers/2 (1 ms)
[----------] 3 tests from All/MouseMoveAzToolsFrameworkTestHelperFixture (4 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (5 ms total)
[  PASSED  ] 3 tests.
```

**Comment by hultonha** [comment]
Okay looking at the platform that failed (Linux) it most likely is the case that `QCursor::pos()` is returning an uninitialized value and is not working on that platform (at least in the test configuration). One option might be to use the overload that accepts a `QScreen` and create one of those in the fixture. Can you give that a try and we can try a new build? Thanks and apologies this is proving to be a little bit of a pain, we'll get there eventually :+1:

**Comment by pollend** [comment]
> Okay looking at the platform that failed (Linux) it most likely is the case that `QCursor::pos()` is returning an uninitialized value and is not working on that platform (at least in the test configuration). One option might be to use the overload that accepts a `QScreen` and create one of those in the fixture. Can you give that a try and we can try a new build? Thanks and apologies this is proving to be a little bit of a pain, we'll get there eventually +1

how do you create a QScreen?

**Comment by hultonha** [comment]
> > Okay looking at the platform that failed (Linux) it most likely is the case that `QCursor::pos()` is returning an uninitialized value and is not working on that platform (at least in the test configuration). One option might be to use the overload that accepts a `QScreen` and create one of those in the fixture. Can you give that a try and we can try a new build? Thanks and apologies this is proving to be a little bit of a pain, we'll get there eventually +1
> 
> how do you create a QScreen?

I looked into it and that seems like a pain 😖 There's an accessor on `qApp` for `primaryScreen()` but no setter - my guess is on Linux it's nullptr. It seems like we don't rely on `QCursor::pos()` in other tests which is why they're not failing. I think it would be better to maintain our own version of _where_ the cursor should be without asking Qt for it. That might be the best solution.

**Comment by hultonha** [comment]
> > > Okay looking at the platform that failed (Linux) it most likely is the case that `QCursor::pos()` is returning an uninitialized value and is not working on that platform (at least in the test configuration). One option might be to use the overload that accepts a `QScreen` and create one of those in the fixture. Can you give that a try and we can try a new build? Thanks and apologies this is proving to be a little bit of a pain, we'll get there eventually +1
> > 
> > 
> > how do you create a QScreen?
> 
> I looked into it and that seems like a pain 😖 There's an accessor on `qApp` for `primaryScreen()` but no setter - my guess is on Linux it's nullptr. It seems like we don't rely on `QCursor::pos()` in other tests which is why they're not failing. I think it would be better to maintain our own version of _where_ the cursor should be without asking Qt for it. That might be the best solution.

One way to do that might be setting up a listener for the mouseMove event (that we could store) when it happens in `MouseMove`? And then use that with `m_rootWidget->mapFromGlobal` instead of `QCursor::pos()`?

File: Code/Framework/AzToolsFramework/AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.cpp [hunk 2]
```diff
@@ -45,15 +44,17 @@ namespace UnitTest
...
         // ^1 To ensure a mouse move event is fired we must call the test mouse move function
         // and also send a mouse move event that matches. Each on their own do not appear to
         // work - please see the links above for more context.
-        QTest::mouseMove(widget, nextPosition);
+        QTest::mouseMove(widget, nextLocalPosition);
         QMouseEvent mouseMoveEvent(
-            QEvent::MouseMove, QPointF(nextPosition), QPointF(nextPosition), Qt::NoButton, mouseButton, Qt::NoModifier);
+            QEvent::MouseMove, QPointF(nextLocalPosition), QPointF(nextGlobalPosition), Qt::NoButton, mouseButton, Qt::NoModifier);
         QApplication::sendEvent(widget, &mouseMoveEvent);
+        QApplication::processEvents();
```
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.cpp:57]: question: Are you sure this is needed?

Comment (pollend) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UnitTest/AzToolsFrameworkTestHelpers.cpp:57]: I don't think so. I was trying things.

**Comment by pollend** [comment]
> > > > Okay looking at the platform that failed (Linux) it most likely is the case that `QCursor::pos()` is returning an uninitialized value and is not working on that platform (at least in the test configuration). One option might be to use the overload that accepts a `QScreen` and create one of those in the fixture. Can you give that a try and we can try a new build? Thanks and apologies this is proving to be a little bit of a pain, we'll get there eventually +1
> > > 
> > > 
> > > how do you create a QScreen?
> > 
> > 
> > I looked into it and that seems like a pain confounded There's an accessor on `qApp` for `primaryScreen()` but no setter - my guess is on Linux it's nullptr. It seems like we don't rely on `QCursor::pos()` in other tests which is why they're not failing. I think it would be better to maintain our own version of _where_ the cursor should be without asking Qt for it. That might be the best solution.
> 
> One way to do that might be setting up a listener for the mouseMove event (that we could store) when it happens in `MouseMove`? And then use that with `m_rootWidget->mapFromGlobal` instead of `QCursor::pos()`?

there is a lot of logic that relies on QCursor to set the position of the cursor. does that event get emitted when you set the position of the QCursor?

**Comment by hultonha** [comment]
> > > > > Okay looking at the platform that failed (Linux) it most likely is the case that `QCursor::pos()` is returning an uninitialized value and is not working on that platform (at least in the test configuration). One option might be to use the overload that accepts a `QScreen` and create one of those in the fixture. Can you give that a try and we can try a new build? Thanks and apologies this is proving to be a little bit of a pain, we'll get there eventually +1
> > > > 
> > > > 
> > > > how do you create a QScreen?
> > > 
> > > 
> > > I looked into it and that seems like a pain confounded There's an accessor on `qApp` for `primaryScreen()` but no setter - my guess is on Linux it's nullptr. It seems like we don't rely on `QCursor::pos()` in other tests which is why they're not failing. I think it would be better to maintain our own version of _where_ the cursor should be without asking Qt for it. That might be the best solution.
> > 
> > 
> > One way to do that might be setting up a listener for the mouseMove event (that we could store) when it happens in `MouseMove`? And then use that with `m_rootWidget->mapFromGlobal` instead of `QCursor::pos()`?
> 
> there is a lot of logic that relies on QCursor to set the position of the cursor. does that event get emitted when you set the position of the QCursor?

I've got an idea, just trying something, will report back shortly 👍

**Comment by hultonha** [comment]
Okay I got it! 😅 

So this is all super hacky right now, I'll tidy it up for tomorrow and there's a few gotchas but it more or less works... 

I have a new type called `MouseMoveDetector` that looks something like this:

```c++
class MouseMoveDetector : public QWidget
{
    Q_OBJECT
public:
    MouseMoveDetector(QWidget* parent = nullptr);
    ~MouseMoveDetector();

    bool eventFilter([[maybe_unused]] QObject* watched, QEvent* event) override;
    //void mouseMoveEvent(QMouseEvent* event);

    QPoint m_mouseGlobalPosition;
};
```

Right now it's a `QWidget` but I'm going to switch it back to a `QObject` and allow the `installEventFilter` to accept an arbitrary `QObject` (or `QWidget`).

`eventFilter` looks like this

```c++
bool MouseMoveDetector::eventFilter([[maybe_unused]] QObject* watched, QEvent* event)
{
    if (const auto eventType = event->type();
        eventType == QEvent::Type::MouseMove)
    {
        auto mouseEvent = static_cast<QMouseEvent*>(event);
        m_mouseGlobalPosition = mouseEvent->globalPos();
    }

    return QWidget::eventFilter(watched, event);
}
```

And then we use it in the test like so...

```
// setup
const MouseMoveParams mouseMoveParams = GetParam();
m_rootWidget->move(mouseMoveParams.widgetPosition);
m_rootWidget->setFixedSize(mouseMoveParams.widgetSize);

MouseMove(m_rootWidget.get(), mouseMoveParams.localCursorPosition, mouseMoveParams.deltaPosition);

// line that replaces QCursor::pos() call
const QPoint mouseLocalPos = m_rootWidget->mapFromGlobal(m_rootWidget->m_mouseGlobalPosition);
```

Right now I just swapped out `QWidget` for `m_rootWidget` with `MouseMoveDetector` (as right now it's a widget) but I'm going to change that as mentioned above. This causes the test to pass and should fail if I revert the code in the `MouseMove` function. I'll paste a neater update tomorrow (I've got lots of random code sprinkled around at the moment). If you can get it working using that approach please feel free to go ahead of I'll add the relevant snippets/updates/diffs here tomorrow (have to run now sorry).

Let me know what you think! Thanks for bearing with me 👍

**Comment by pollend** [comment]
> Okay I got it! sweat_smile
> 
> So this is all super hacky right now, I'll tidy it up for tomorrow and there's a few gotchas but it more or less works...
> 
> I have a new type called `MouseMoveDetector` that looks something like this:
> 
> ```c++
> class MouseMoveDetector : public QWidget
> {
>     Q_OBJECT
> public:
>     MouseMoveDetector(QWidget* parent = nullptr);
>     ~MouseMoveDetector();
> 
>     bool eventFilter([[maybe_unused]] QObject* watched, QEvent* event) override;
> 
>     QPoint m_mouseGlobalPosition;
> };
> ```
> 
> Right now it's a `QWidget` but I'm going to switch it back to a `QObject` and allow the `installEventFilter` to accept an arbitrary `QObject` (or `QWidget`).
> 
> `eventFilter` looks like this
> 
> ```c++
> bool MouseMoveDetector::eventFilter([[maybe_unused]] QObject* watched, QEvent* event)
> {
>     if (const auto eventType = event->type();
>         eventType == QEvent::Type::MouseMove)
>     {
>         auto mouseEvent = static_cast<QMouseEvent*>(event);
>         m_mouseGlobalPosition = mouseEvent->globalPos();
>     }
> 
>     return QWidget::eventFilter(watched, event);
> }
> ```
> 
> And then we use it in the test like so...
> 
> ```
> // setup
> const MouseMoveParams mouseMoveParams = GetParam();
> m_rootWidget->move(mouseMoveParams.widgetPosition);
> m_rootWidget->setFixedSize(mouseMoveParams.widgetSize);
> 
> MouseMove(m_rootWidget.get(), mouseMoveParams.localCursorPosition, mouseMoveParams.deltaPosition);
> 
> // line that replaces QCursor::pos() call
> const QPoint mouseLocalPos = m_rootWidget->mapFromGlobal(m_rootWidget->m_mouseGlobalPosition);
> ```
> 
> Right now I just swapped out `QWidget` for `m_rootWidget` with `MouseMoveDetector` (as right now it's a widget) but I'm going to change that as mentioned above. This causes the test to pass and should fail if I revert the code in the `MouseMove` function. I'll paste a neater update tomorrow (I've got lots of random code sprinkled around at the moment). If you can get it working using that approach please feel free to go ahead of I'll add the relevant snippets/updates/diffs here tomorrow (have to run now sorry).
> 
> Let me know what you think! Thanks for bearing with me +1

feel free to just commit to this PR.

**Comment by hultonha** [comment]
@pollend I have my updates ready to push but I'm not sure how to push my commit to this branch/PR - is this something you can do using the `gh` cli? I had a look [here ](https://cli.github.com/manual/) but couldn't see it. Thanks!

**Comment by hultonha** [comment]
> @pollend I have my updates ready to push but I'm not sure how to push my commit to this branch/PR - is this something you can do using the `gh` cli? I had a look [here ](https://cli.github.com/manual/) but couldn't see it. Thanks!

I've opened this PR for now - https://github.com/o3de/o3de/pull/6725. I can close it and update this one if you let me know how or we can close this one and use the other one, I'm happy either way. Thanks!

**Comment by hultonha** [comment]
Managed to update with my latest commit, kicked of a build now, fingers crossed it works! 😝 🤞

**Comment by pollend** [comment]
Looks like it failed on something unrelated so I just need to rebase

**Comment by pollend** [comment]
> Managed to update with my latest commit, kicked of a build now, fingers crossed it works! stuck_out_tongue_closed_eyes crossed_fingers

I was just looking. umm seems like github cli can't push to a PR, I guess I just assumed it could do it.

**Comment by lawsonamzn** [comment]
> > Managed to update with my latest commit, kicked of a build now, fingers crossed it works! stuck_out_tongue_closed_eyes crossed_fingers
> 
> I was just looking. umm seems like github cli can't push to a PR, I guess I just assumed it could do it.

Usually all you have to do is push to the branch in origin its based off on (ie, update the branch the PR is made from at the source), and github will automatically update the PR.

**Comment by hultonha** [comment]
> > > Managed to update with my latest commit, kicked of a build now, fingers crossed it works! stuck_out_tongue_closed_eyes crossed_fingers
> > 
> > 
> > I was just looking. umm seems like github cli can't push to a PR, I guess I just assumed it could do it.
> 
> Usually all you have to do is push to the branch in origin its based off on (ie, update the branch the PR is made from at the source), and github will automatically update the PR.

That's right I wound up just doing `git push pollend bugfix/6481-fix-mouse-move-utility` and it worked out okay 👍

**Comment by hultonha** [comment]
> > Managed to update with my latest commit, kicked of a build now, fingers crossed it works! stuck_out_tongue_closed_eyes crossed_fingers
> 
> I was just looking. umm seems like github cli can't push to a PR, I guess I just assumed it could do it.

Ah yes I saw that 😞 If you could rebase or merge development into this branch again, whichever you prefer, then I can kick off another build 👍 Thanks!

**Comment by pollend** [comment]
> > > Managed to update with my latest commit, kicked of a build now, fingers crossed it works! stuck_out_tongue_closed_eyes crossed_fingers
> > 
> > 
> > I was just looking. umm seems like github cli can't push to a PR, I guess I just assumed it could do it.
> 
> Ah yes I saw that disappointed If you could rebase or merge development into this branch again, whichever you prefer, then I can kick off another build +1 Thanks!

I'll have to do it later tonight

**Comment by hultonha** [comment]
@pollend Oh no I think that force push just blew away my commit 😩 As soon as we have two people working on a branch we should avoid rebasing and just merge to be safe. I'll see if I can redo my commit again

**Comment by hultonha** [comment]
Managed to keep the commit as a patch and reapply it (also renamed the new test file slightly to match the convention of the file it's testing). Hopefully this next build succeeds! 😬 🤞

**Comment by hultonha** [comment]
It passed! 😄 🎉 I'm totally merging this now before anything else happens 😅

**Comment by hultonha** [comment]
Thank you for your patience and persistent @pollend, we got there in the end! 🙂

**Comment by pollend** [comment]
Need to fix the other pr

**Comment by hultonha** [comment]
> Need to fix the other pr

Yup that hack/workaround can now be removed 👍

**Pull request merged**